File: app/src/main/java/org/tasks/caldav/CaldavAccountSettingsActivity.java
package org.tasks.caldav;

import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;
import com.todoroo.astrid.helper.UUIDHelper;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.data.CaldavAccount;
import org.tasks.injection.ActivityComponent;
import timber.log.Timber;

public class CaldavAccountSettingsActivity extends BaseCaldavAccountSettingsActivity
    implements Toolbar.OnMenuItemClickListener {

  @Inject CaldavClient client;

  private AddCaldavAccountViewModel addCaldavAccountViewModel;
  private UpdateCaldavAccountViewModel updateCaldavAccountViewModel;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    ViewModelProvider provider = new ViewModelProvider(this);
    addCaldavAccountViewModel = provider.get(AddCaldavAccountViewModel.class);
    updateCaldavAccountViewModel = provider.get(UpdateCaldavAccountViewModel.class);

    addCaldavAccountViewModel.observe(this, this::addAccount, this::requestFailed);
    updateCaldavAccountViewModel.observe(this, this::updateAccount, this::requestFailed);
  }

  @Override
  protected int getDescription() {
    return R.string.caldav_account_description;
  }

  private void addAccount(String principal) {
    hideProgressIndicator();

    Timber.d("Found principal: %s", principal);

    CaldavAccount newAccount = new CaldavAccount();
    newAccount.setName(getNewName());
    newAccount.setUrl(principal);
    newAccount.setUsername(getNewUsername());
    newAccount.setPassword(encryption.encrypt(getNewPassword()));
    newAccount.setUuid(UUIDHelper.newUUID());
    newAccount.setId(caldavDao.insert(newAccount));

    setResult(RESULT_OK);
    finish();
  }

  private void updateAccount(String principal) {
    hideProgressIndicator();

    caldavAccount.setName(getNewName());
    caldavAccount.setUrl(principal);
    caldavAccount.setUsername(getNewUsername());
    caldavAccount.setError("");
    if (passwordChanged()) {
      caldavAccount.setPassword(encryption.encrypt(getNewPassword()));
    }
    caldavAccount.setSuppressRepeatingTasks(binding.repeat.isChecked());
    caldavDao.update(caldavAccount);

    setResult(RESULT_OK);
    finish();
  }

  @Override
  protected void addAccount(String url, String username, String password) {
    addCaldavAccountViewModel.addAccount(client, url, username, password);
  }

  @Override
  protected void updateAccount(String url, String username, String password) {
    updateCaldavAccountViewModel.updateCaldavAccount(client, url, username, password);
  }

  @Override
  protected void updateAccount() {
    updateAccount(caldavAccount.getUrl());
  }

  @Override
  protected String getNewPassword() {
    String input = binding.password.getText().toString().trim();
    return PASSWORD_MASK.equals(input) ? encryption.decrypt(caldavAccount.getPassword()) : input;
  }

  @Override
  protected String getHelpUrl() {
    return "https://tasks.org/caldav";
  }

  @Override
  public void inject(ActivityComponent component) {
    component.inject(this);
  }
}

File: app/src/main/java/org/tasks/caldav/CaldavAccountSettingsActivity.kt
package org.tasks.caldav

import android.app.Activity
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.widget.Toolbar
import androidx.core.view.isVisible
import org.tasks.data.UUIDHelper
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.R
import org.tasks.analytics.Constants
import org.tasks.data.entity.CaldavAccount
import timber.log.Timber

@AndroidEntryPoint
class CaldavAccountSettingsActivity : BaseCaldavAccountSettingsActivity(), Toolbar.OnMenuItemClickListener {
    private val viewModel: CaldavAccountViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel.inFlight.observe(this) { binding.progressBar.progressBar.isVisible = it }
        viewModel.error.observe(this) { throwable ->
            throwable?.let {
                requestFailed(it)
                viewModel.error.value = null
            }
        }
        viewModel.finish.observe(this) {
            setResult(RESULT_OK, it)
            finish()
        }
    }

    override val description: Int
        get() = R.string.caldav_account_description

    private suspend fun addAccount(principal: String) {
        hideProgressIndicator()
        Timber.d("Found principal: %s", principal)
        caldavDao.insert(
            CaldavAccount(
                name = newName,
                url = principal,
                username = newUsername,
                password = encryption.encrypt(newPassword!!),
                uuid = UUIDHelper.newUUID(),
            )
        )
        firebase.logEvent(
                R.string.event_sync_add_account,
                R.string.param_type to Constants.SYNC_TYPE_CALDAV
        )
        setResult(Activity.RESULT_OK)
        finish()
    }

    private suspend fun updateAccount(principal: String?) {
        hideProgressIndicator()
        caldavAccount!!.name = newName
        caldavAccount!!.url = principal
        caldavAccount!!.username = newUsername
        caldavAccount!!.error = ""
        caldavAccount!!.serverType = serverType.value
        if (passwordChanged()) {
            caldavAccount!!.password = encryption.encrypt(newPassword!!)
        }
        caldavDao.update(caldavAccount!!)
        setResult(Activity.RESULT_OK)
        finish()
    }

    override suspend fun addAccount(url: String, username: String, password: String) {
        viewModel.addAccount(url, username, password)?.let { addAccount(it) }
    }

    override suspend fun updateAccount(url: String, username: String, password: String) {
        viewModel.updateCaldavAccount(url, username, password)?.let { updateAccount(it) }
    }

    override suspend fun updateAccount() = updateAccount(caldavAccount!!.url)

    override val newPassword: String?
        get() {
            val input = binding.password.text.toString().trim { it <= ' ' }
            return if (PASSWORD_MASK == input) encryption.decrypt(caldavAccount!!.password) else input
        }

    override val helpUrl = R.string.url_caldav
}